Add custom decredmaterial button and clickable

beansgum opened the pull request:
Closes #592
This pr implements a custom clickable highlight that allows custom color and border radius properties. A custom button layout is also added, it makes use of a custom clickable highlight and allows switching to a disabled state without directly changing the colors.

beansgum committed:
Add custom button layout

- Remove all direct accesses of ButtonStyle clickable

beansgum committed:
button: disable user interactions if disabled

beansgum committed:
Add support for custom button highlight color

- add predefined styles for outline button

beansgum committed:
Replace instances with explicit outline buttons style with predefined style

beansgum committed:
Fix incorrect button background in dropdown & switch button

beansgum committed:
Add clickable layout that supports custom background & radius

- apply custom clickable layout to clickable list

beansgum committed:
Add Hover to decredmaterial Clickable

beansgum committed:
Replace usages of Clickable with decredmaterial.Clickable

beansgum committed:
Correctly handle button disabled states in all pages

beansgum committed:
Wrap create wallet button in a white background

Sirmorrison commented on ui/decredmaterial/button.go:
```diff
@@ -29,14 +40,39 @@ type IconButton struct {
 func (t *Theme) Button(button *widget.Clickable, txt string) Button {
+	buttonStyle := material.Button(t.Base, button, txt)
+	buttonStyle.TextSize = values.TextSize16
+	// buttonStyle.Font.Weight = text.Medium
```
Remove if no longer needed.

Sirmorrison commented on ui/decredmaterial/button.go:
```diff
@@ -6,18 +6,29 @@ import (
+	disabledTextColor  color.NRGBA
+	HighlightColor     color.NRGBA
+
+	// Padding layout.Inset
```
Remove if no longer needed.

Sirmorrison commented on ui/page/components/nav_drawer.go:
```diff
@@ -55,49 +55,49 @@ func (nd *NavDrawer) LayoutNavDrawer(gtx layout.Context) layout.Dimensions {
+					// 	background = nd.Theme.Color.ActiveGray
+					// }
+
+					// return nd.layoutNavRow(gtx, background, nd.DrawerNavItems[i].Clickable, func(gtx C) D {
```
Remove if no longer needed.

Sirmorrison commented on ui/page/components/nav_drawer.go:
```diff
@@ -55,49 +55,49 @@ func (nd *NavDrawer) LayoutNavDrawer(gtx layout.Context) layout.Dimensions {
+							}),
+						)
 					})
+					// })
```
Remove if no longer needed.

Sirmorrison commented on ui/page/components/nav_drawer.go:
```diff
@@ -119,41 +119,41 @@ func (nd *NavDrawer) LayoutTopBar(gtx layout.Context) layout.Dimensions {
+				// background := nd.Theme.Color.Surface
+				// if nd.AppBarNavItems[i].PageID == nd.CurrentPage {
+				// 	background = nd.Theme.Color.ActiveGray
+				// }
```
Remove if no longer needed.

Sirmorrison commented on ui/page/components/nav_drawer.go:
```diff
@@ -119,41 +119,41 @@ func (nd *NavDrawer) LayoutTopBar(gtx layout.Context) layout.Dimensions {
-									return layout.Inset{
-										Left: values.MarginPadding0,
-									}.Layout(gtx, func(gtx C) D {
+				// return nd.layoutNavRow(gtx, background, nd.AppBarNavItems[i].Clickable, func(gtx C) D {
```
Remove if no longer needed.

Sirmorrison commented on ui/page/components/nav_drawer.go:
```diff
@@ -119,41 +119,41 @@ func (nd *NavDrawer) LayoutTopBar(gtx layout.Context) layout.Dimensions {
+						)
 					})
 				})
+				// })
```
Remove if no longer needed.

Sirmorrison reviewed (commented):
We will need to open an upstream issue to gio for the clickable widget drawInk(), as the issue with the clickable widget comes from there.

beansgum committed:
Fix nav layout

beansgum committed:
Add clickable hover to account selector modal

beansgum committed:
Add clickable hover to dropdown widget

dreacot reviewed (changes requested):
The width of the selector on the dropdown should be full width there should be no margin on the selector also
<img width="246" alt="Screenshot 2021-09-16 at 8 29 00 AM" src="https://user-images.githubusercontent.com/25265396/133569642-a8a34575-8af0-479a-8ecb-dde09ebc034a.png">

dreacot reviewed (changes requested):
The hover effect on the account selector should have a padding
<img width="467" alt="Screenshot 2021-09-16 at 8 34 17 AM" src="https://user-images.githubusercontent.com/25265396/133570145-459057aa-9526-47aa-9814-14281580bfd8.png">

dreacot commented on ui/decredmaterial/dropdown.go:
```diff
@@ -164,24 +164,28 @@ func (c *DropDown) layoutActiveIcon(index int, isFirstOption bool) layout.FlexCh
 
 func (c *DropDown) layoutOption(gtx C, itemIndex int, isFirstOption bool) D {
-	btn := c.items[itemIndex].button
+	// btn := c.items[itemIndex].clickable
```
remove comment if not necessary

dreacot commented on ui/decredmaterial/dropdown.go:
```diff
@@ -164,24 +164,28 @@ func (c *DropDown) layoutActiveIcon(index int, isFirstOption bool) layout.FlexCh
+
+	// 	}),
+	// 	layout.Expanded(btn.Layout),
+	// )
```
remove comment if not necessary

dreacot commented on ui/page/components/account_selector.go:
```diff
@@ -402,65 +397,64 @@ func (asm *AccountSelectorModal) Layout(gtx layout.Context) layout.Dimensions {
-	pointer.Rect(image.Rectangle{Max: gtx.Constraints.Max}).Add(gtx.Ops)
-	account.clickEvent.Add(gtx.Ops)
+	// pointer.Rect(image.Rectangle{Max: gtx.Constraints.Max}).Add(gtx.Ops)
+	// account.clickEvent.Add(gtx.Ops)
```
remove comment if not necessary

beansgum committed:
Remove unused comments

Sirmorrison reviewed (changes requested):
* When the side and top nav buttons are hovered, the hover effect do not display.
* Also, on the dropdown, the clickable even behaves unexpectedly as i would have to click twice for the dropdown to be displayed.
* The create wallet button at the bottom right on the wallet page does not have the hover or click hightlight effect anymore

![ezgif com-gif-maker(8)](https://user-images.githubusercontent.com/27733432/133593872-d1d1a2f3-42f3-4925-b971-d7fc1174d8df.gif)

* When the loader icon is displayed, the cancel button should be hidden. 

<img width="584" alt="image" src="https://user-images.githubusercontent.com/27733432/133593823-2ae93e27-392f-4a7d-bb15-e1db4360d26c.png">

My recommendation will be to have this PR split into 2 different PRs.
* The first to resolve the #595 
* The second to resolve clickable events #584 

This will enable us properly test the custom clickable widget and not block issue #595.

beansgum committed:
Add clickable to linear layout

- Fix nav drawer clickable layout

beansgum committed:
Remove clickable parameter from outline button

beansgum committed:
Remove clickable parameter from all buttons

beansgum committed:
Fix dropdown layout clickable

beansgum committed:
Cleanup dropdown widget code

beansgum committed:
Add padding to account selector rows

dreacot commented on ui/decredmaterial/dropdown.go:
```diff
@@ -21,127 +22,116 @@ type DropDown struct {
+				break
+				// if len(d.items[index].Text) > 12 {
+				// 	txt = d.items[index].Text[:12] + "..."
+				// }
```
remove comment

beansgum committed:
Fix highlight colors

beansgum committed:
Cleanup more page row layout

Sirmorrison commented on ui/decredmaterial/dropdown.go:
```diff
@@ -21,127 +22,114 @@ type DropDown struct {
-			for c.items[index].button.Clicked() {
-				c.selectedIndex = index
-				txt := c.items[index].Text
-				if len(c.items[index].Text) > 12 {
-					txt = c.items[index].Text[:12] + "..."
-				}
-				c.items[0].label.Text = txt
```
This lines of code help truncate the wallet name should incase the wallet name is too long. Currently, we dont have a limit to the length of the wallet name, and to prevent the wallet name dropdown from being too long, it was given a maximum width.

This is the effect of removing the code.
![image](https://user-images.githubusercontent.com/27733432/134657456-aa7ee058-46b3-465d-a4bf-b19a68312d1d.png)

Sirmorrison reviewed (changes requested):
![image](https://user-images.githubusercontent.com/27733432/134658174-ffc94936-c45f-470d-9a09-752e3cf29d01.png)

Also, lets remove the top and bottom padding.
I understand you mentioned its from the design, however, the design did not implement the hover feature.
The padding is not nice on the UI so lets remove it.

This was what we had.
![image](https://user-images.githubusercontent.com/27733432/134660071-2309844e-a3de-4602-b0e4-9cc294e92ffc.png)

beansgum committed:
Truncate wallet name

beansgum committed:
Fix reconnect button bug

- Reduce tx details copy button text size and remove redundant creation
  of copy buttons.
- Fix clickable highlight bug

beansgum committed:
Fix click highlight color

beansgum committed:
Fix wallet button highlight

Sirmorrison reviewed (commented):
On the more page there is this extra shadow which is not present on master.
<img width="795" alt="image" src="https://user-images.githubusercontent.com/27733432/135037419-593b1acd-c36e-4570-885c-07b7c6b7039a.png">

<img width="795" alt="image" src="https://user-images.githubusercontent.com/27733432/135037572-8ad457a0-5495-4269-889a-b540560b9dda.png">

beansgum committed:
Merge branch 'master' into action_buttons
